README.md
# pyQUEST

Count unique reads and optionally match them to a given library (exact matching only).

Input files:

- SAM/BAM/CRAM/FASTQ file
- [library file](#library) (library-dependent mode only)

Output files:

- **library-independent count**:
  - [counts](#library-independent-counts)
  - [statistics](#library-independent-stats-file)
- **library-dependent count**:
  - [counts](#library-dependent-counts)
  - [statistics](#library-dependent-stats-file)

Notes:

- only supports single-sample input files
- reads with ambiguous nucleotides are discarded
- masked reads are discarded

## Setup

Using a Python virtual environment:

```sh
python -m venv .venv
source .venv/bin/activate
pip install -r requirements.txt
pip install .
```

The Docker image can be built as follows:

```sh
docker build -t pyquest .
```

## Usage

```
Usage: pyquest [OPTIONS] QUERIES

  Count reads and optionally map them to a library.

  QUERIES: Query sequence file (fastq[.gz], sam, bam, cram)

Options:
  -o, --output PATH               Final output to this filename prefix
                                  [required]
  --min-length INTEGER RANGE      Minimum read length  [default: 1; x>=1]
  --most-common INTEGER RANGE     Output top X most common unique read
                                  sequences in FASTA format  [1<=x<=50]

Input sample metadata:         Options adding information to the input
    -s, --sample TEXT             Sample name to apply to count column,
                                  required for fastq, interrogate header for
                                  others when not defined.
    -r, --reference FILE          Required for CRAM

Library-dependent:             Options specific to library-dependent
                                  counting
    -l, --library FILE            Expanded library definition TSV file with
                                  optional headers (common format for
                                  single/dual/other)
    --low-count INTEGER RANGE     *.stats.json includes
                                  low_count_guides_lt_{15,30}, this option
                                  allow specification of an additional cut-
                                  off.  [x>=0]

Performance:                   Options to tune the performance
    -c, --cpus INTEGER RANGE      CPUs to use (0 to detect)  [default: 1;
                                  x>=0]

Debug:                         Options specific to troubleshooting, testing
                                  and debugging
    --loglevel [WARNING|INFO|DEBUG]
                                  Set logging verbosity  [default: INFO]
    --no-compression              Disable output compression
  --version                       Show the version and exit.
  --help                          Show this message and exit.
```

With Docker:

```sh
# Output in the current directory
mkdir -p output
docker run \
    -v "$PWD/test.queries.bam":/tmp/x.bam:ro \
    -v "$PWD/output":/output \
    pyquest \
        pyquest \
            -o /output/something \
            --sample XYZ \
            --no-compression \
            /tmp/x.bam
```

## File header formats

TSV headers may contain metadata in the form of key-value pairs thus formatted:

```
##<KEY>: <VALUE>
```

The column headers, separated by tabs, immediately follow the metadata lines and are preceded by a single `#` character, *e.g.*:

```
#<FIELD 1>	<FIELD 2>	<FIELD 3>
```

### Count header

|Field|Format|Description|
|-|-|-|
|`Command`|string|Full command|
|`Version`|`x.y.z`|Tool version|

### Library header

Currently, ignored.

## File formats

### Library

**Format**: TSV with [library header](#library-header)

The headers are ignored, and therefore the relevant fields are identified by their position. Here we indicate the field positions as one-based, with their corresponding field names in the [library-dependent counts](#library-dependent-counts).

|Position|Counts field|Format|Description|
|-|-|-|-|
|1|`ID`|string|Library sequence identifier|
|2|`NAME`|string|Library sequence name|
|3|`SEQUENCE`|`[ACGT]+`|DNA sequence|

*E.g.*:

```
## ...
# ...
1	some-name-1	AAAAAAAAATCCAGAACCT
2	some-name-2	AAAAAAATATGCCCGTGGA
3	some-name-3	AAAAAAGCATTTAGGCAGG
4	some-name-4	AAAAAAGCTTGCATTAGAC
5	some-name-5	AAAAAATATCGTGTCAAGT
6	some-name-6	AAAAAATCAGCCACGCGAC
```

### Library-independent counts

**Format**: TSV with [count header](#count-header) (gzip'ed by default)

|Field|Format|Description|
|-|-|-|
|`SEQUENCE`|`[ACGT]+`|Unique DNA sequence|
|`LENGTH`|integer|Length of the sequence|
|`COUNT`|integer|Number of reads|

*E.g.*:

```
##Command: pyquest -o output --min-length 0 --low-count 2 -l guides.tsv --sample XYZ --no-compression test.queries.bam
##Version: 1.0.0
#SEQUENCE	LENGTH	COUNT
AAAAAAGCTTGCATTAGAC	19	25
AAAAAATATCGTGTCAAGT	19	26
AAAAAATGTCAGTCGAGTG	19	34
AAAAACAAGCGCACCACCG	19	1
AAAAACACTTCCATGCAAA	19	25
AAAAACGTATTTAGCCGAA	19	23
```

### Library-dependent counts

**Format**: TSV with [count header](#count-header) (gzip'ed by default)

|Field|Format|Description|
|-|-|-|
|`ID`|string|Library sequence identifier|
|`NAME`|string|Library sequence name|
|`SEQUENCE`|`[ACGT]+`|DNA sequence|
|`LENGTH`|integer|Length of the DNA sequence|
|`COUNT`|integer|Number of reads|
|`UNIQUE`|0\|1|Whether the sequence is unique in the library|
|`SAMPLE`|string|Name of the sample of origin of the reads|

*E.g.*:

```
##Command: pyquest -o output --min-length 0 --low-count 2 -l guides.tsv --sample XYZ --no-compression test.queries.bam
##Version: 1.0.0
#ID	NAME	SEQUENCE	COUNT	UNIQUE	SAMPLE
1	some-name-1	AAAAAAAAATCCAGAACCT	0	1	XYZ
2	some-name-2	AAAAAAATATGCCCGTGGA	0	1	XYZ
3	some-name-3	AAAAAAGCATTTAGGCAGG	0	1	XYZ
4	some-name-4	AAAAAAGCTTGCATTAGAC	25	1	XYZ
5	some-name-5	AAAAAATATCGTGTCAAGT	26	1	XYZ
6	some-name-6	AAAAAATCAGCCACGCGAC	0	1	XYZ
```

### Library-independent stats file

**Format**: JSON

|Field|Format|Description|
|-|-|-|
|`sample_name`|string|Name of the sample|
|`input_reads`|integer|Total input reads|
|`total_reads`|integer|Total reads passed on to counting|
|`discarded_reads`|integer|Total reads discarded before counting|
|`vendor_failed_reads`|integer|Total reads with the `QCFAIL` flag|
|`length_excluded_reads`|integer|Total reads discarded because shorter than a user-defined threshold|
|`ambiguous_nt_reads`|integer|Total reads with ambiguous nucleotides|
|`masked_reads`|integer|Total soft-masked reads|
|`zero_length_reads`|integer|Total zero-length reads|

*E.g.*:

```json
{
    "version": "1.0.0",
    "command": "pyquest -o output --min-length 0 --low-count 2 -l guides.tsv --sample XYZ --no-compression test.queries.bam",
    "sample_name": "XYZ",
    "total_reads": 1020769,
    "vendor_failed_reads": 0,
    "length_excluded_reads": 0,
    "ambiguous_nt_reads": 0,
    "masked_reads": 0
}
```

### Library-dependent stats file

**Format**: JSON

The library-dependent count statistics include the [library-dependent count statistics](#library-independent-stats-file).

All statistics are computed on the read counts of unique targets, excluding those discarded based on their length. The number of low count templates (`zero_count_templates` and `low_count_templates_*`) also excludes the targets with short sequences.

|Field|Format|Description|
|-|-|-|
|`mapped_to_template_reads`|integer|Total reads mapping to the library|
|`mean_count_per_template`|decimal|Mean reads per template|
|`median_count_per_template`|decimal|Median reads per template|
|`multimap_reads`|integer|Total reads mapping to more than one template|
|`unmapped_reads`|integer|Total reads mapping to no template|
|`total_templates`|integer|Total number of templates|
|`total_unique_templates`|integer|Total number of unique templates|
|`length_excluded_templates`|integer|Total number of unique templates excluded by length|
|`zero_count_templates`|integer|Total number of unique templates with no reads mapping to them|
|`low_count_templates_lt_15`|integer|Total number of unique templates with less than 15 reads mapping to them|
|`low_count_templates_lt_30`|integer|Total number of unique templates with less than 30 reads mapping to them|
|`low_count_templates_user`|object\|`null`|Total number of unique templates with less than a user-defined number of reads mapping to them (optional)|
|`gini_coefficient`|decimal|Gini coefficient of the mapping read counts|

*E.g.*:

```json
{
    "version": "1.0.0",
    "command": "pyquest -o output --min-length 3 --low-count 2 -l guides.tsv --sample XYZ --no-compression test.queries.sam",
    "sample_name": "XYZ",
    "input_reads": 1020770,
    "total_reads": 1020766,
    "discarded_reads": 4,
    "vendor_failed_reads": 0,
    "length_excluded_reads": 1,
    "ambiguous_nt_reads": 2,
    "masked_reads": 2,
    "mapped_to_template_reads": 1020766,
    "mean_count_per_template": 10.1,
    "median_count_per_template": 0,
    "multimap_reads": 0,
    "unmapped_reads": 0,
    "total_templates": 101064,
    "total_unique_templates": 101064,
    "length_excluded_templates": 0,
    "zero_count_templates": 60927,
    "low_count_templates_lt_15": 72265,
    "low_count_templates_lt_30": 84339,
    "low_count_templates_user": {
      "lt": 2,
      "count": 61744
    },
    "gini_coefficient": 0.73
}
```


CHANGELOG.md
# Changelog

## [1.1.0] - 2023-10-10

Handling of zero-length reads and most common reads fasta output.

## [1.0.0] - 2023-06-22

Initial release.


src/pyquest/main.py
# pyQUEST
#
# Copyright (C) 2023 Genome Research Ltd.
#
# Author: Luca Barbon
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

import logging
import os
import platform
import sys
from typing import NoReturn

import click
from click_option_group import OptionGroup

from . import __version__
from .app_info import AppInfo
from .counting import run_library_independent_counting
from .errors import CustomException, InvalidInputFormatError, InputReadError, InvalidLibraryError, MissingMetadataError, UnsupportedData
from .library import TargetLibrary
from .readers.read_file_info import ReadFileFormat, ReadFileInfo
from .writer import write_stats, write_unique_counts, write_most_common_reads


LOG_LEVELS = [
    'WARNING',
    'INFO',
    'DEBUG'
]

HELP_LIBRARY = "Expanded library definition TSV file with optional headers (common format for single/dual/other)"
HELP_MIN_LENGTH = "Minimum read length"
HELP_MOST_COMMON = "Output top X most common unique read sequences in FASTA format"
HELP_SAMPLE = "Sample name to apply to count column, required for fastq, interrogate header for others when not defined."
HELP_OUTPUT = "Final output to this filename prefix"
HELP_LOW_COUNT = "*.stats.json includes low_count_guides_lt_{15,30}, this option allow specification of an additional cut-off."
HELP_REFERENCE = "Required for CRAM"
HELP_CPUS = "CPUs to use (0 to detect)"


existing_file_path = click.Path(
    exists=True,
    file_okay=True,
    dir_okay=False,
    readable=True,
    resolve_path=True)


class PlatformError(Exception):
    pass


def abort(ex: CustomException) -> NoReturn:
    logging.error(ex.message)
    sys.exit(1)


def get_cpus(cpus: int) -> int:
    if cpus > 0:
        return cpus
    if platform.system() == 'Darwin':
        logging.error("Process affinity detection not available on macOS: please set the CPU count!")
        raise PlatformError
    return len(os.sched_getaffinity(0))  # type: ignore[attr]


def setup_fs(output_dir: str) -> None:
    # TODO: verify file vs. directory as output
    outfolder = os.path.dirname(os.path.abspath(output_dir))
    if not os.path.exists(outfolder):
        os.makedirs(outfolder, exist_ok=True)


lib_dep_opts = OptionGroup("\nLibrary-dependent", help="Options specific to library-dependent counting")
sample_opts = OptionGroup("\nInput sample metadata", help="Options adding information to the input")
perf_opts = OptionGroup("\nPerformance", help="Options to tune the performance")
debug_opts = OptionGroup("\nDebug", help="Options specific to troubleshooting, testing and debugging")


@click.command()
@click.argument('queries', required=True, type=existing_file_path)
@click.option(
    '-o',
    '--output',
    required=True,
    type=click.Path(exists=False, file_okay=True, resolve_path=True),
    help=HELP_OUTPUT
)
@click.option('--min-length', type=click.IntRange(min=1), default=1, help=HELP_MIN_LENGTH, show_default=True)
@click.option('--most-common', type=click.IntRange(min=1, max=50), default=None, help=HELP_MOST_COMMON, show_default=True)
@sample_opts.option('-s', '--sample', type=str, help=HELP_SAMPLE)
@sample_opts.option('-r', '--reference', type=existing_file_path, help=HELP_REFERENCE)
@lib_dep_opts.option('-l', '--library', default=None, type=existing_file_path, help=HELP_LIBRARY)
@lib_dep_opts.option('--low-count', type=click.IntRange(min=0), default=None, help=HELP_LOW_COUNT, show_default=True)
@perf_opts.option('-c', '--cpus', type=click.IntRange(min=0), default=1, show_default=True, help=HELP_CPUS)
@debug_opts.option(
    '--loglevel',
    default='INFO',
    show_default=True,
    type=click.Choice(LOG_LEVELS, case_sensitive=False),
    help="Set logging verbosity"
)
@debug_opts.option(
    '--no-compression',
    is_flag=True,
    default=False,
    help="Disable output compression"
)
@click.version_option(__version__)
def main(
    library: str | None,
    sample: str | None,
    output: str,
    low_count: int | None,
    most_common: int | None,
    min_length: int,
    reference: str | None,
    queries: str,
    cpus: int,
    loglevel: str,
    no_compression: bool
):
    """
    Count reads and optionally map them to a library.

    QUERIES: Query sequence file (fastq[.gz], sam, bam, cram)
    """

    # Setup logger
    logging.basicConfig(
        level=logging._nameToLevel[loglevel.upper()],
        format="%(levelname)s: %(message)s")

    # Warn on ignored options
    if not library and low_count is not None:
        logging.warning("Low count option ignored in library-independent mode.")

    # Get usable CPU's
    try:
        usable_cpus: int = get_cpus(cpus)
    except PlatformError:
        sys.exit(1)

    # Validate sample name
    read_file_info = ReadFileInfo.from_path(queries)
    if read_file_info.fmt == ReadFileFormat.FASTQ and not sample:
        logging.error("When a FASTQ is provided, a sample name is required!")
        sys.exit(1)

    # Setup output directory
    setup_fs(output)

    app_info = AppInfo(
        __version__,
        ' '.join([os.path.basename(sys.argv[0]), *sys.argv[1:]])
    )

    # Library-independent counting
    logging.info("Loading reads...")
    try:
        unique_count, stats, query_counts = run_library_independent_counting(
            read_file_info, sample, usable_cpus, reference, min_length=min_length)
    except (InvalidInputFormatError, MissingMetadataError) as ex:
        abort(ex)

    if stats.total_zero_reads == stats.input_reads:
        # Error if all reads are zero-length
        abort(InputReadError("All reads zero-length!"))
    elif stats.total_zero_reads > 0:
        # Else warn user of number of 0-length reads
        logging.warning(f"{stats.total_zero_reads} zero-length reads were found")

    # Write unique counts to file
    logging.info("Writing library-independent counts...")
    write_unique_counts(query_counts, app_info, output, compress=not no_compression)

    # Write X most common unique reads to file if specified
    if most_common is not None:
        write_most_common_reads(most_common, query_counts, output, compress=not no_compression)

    if library:

        # Load library
        # TODO: verify stat's are not affected by pre-filtering the targets by length
        logging.info("Loading library...")
        try:
            targets = TargetLibrary.load(library, min_length=min_length)
        except InvalidLibraryError as ex:
            abort(ex)

        logging.info("Finding exact matches and writing library-dependent counts...")
        fp: str = f"{output}.lib_counts.tsv"
        if not no_compression:
            fp += '.gz'
        try:
            stats = targets.map_and_write(
                app_info,
                stats,
                query_counts,
                fp,
                compress=not no_compression,
                custom_count_threshold=low_count)
        except UnsupportedData as ex:
            abort(ex)

    # Write stats to file
    logging.info("Writing stats report...")
    stats_fp: str = f"{output}.stats.json"
    write_stats(app_info, stats, stats_fp)
